src/cat.jl

```diff
@@ -0,0 +1,157 @@
+# separate out in module to have the abliity of defining cat
+"""
+    module Concatenate
+
+Alternative implementation for `Base.cat` through [`concatenate(!)`](@ref).
+
+This is mostly a copy of the Base implementation, with the main difference being
+that the destination is chosen based on all inputs instead of just the first.
+
+Additionally, we have an intermediate representation in terms of a Concatenated object,
+reminiscent of how Broadcast works.
+
+The various entry points for specializing behavior are:
+
+* Destination selection can be achieved through
+
+    Base.similar(cat::Concatenated{Interface}, ::Type{T}, axes) where {Interface}
+
+* Implementation for moving one or more arguments into the destionation through
+
+    copy_offset!(dest, shape, catdims, offsets, args...)
+    copy_offset1!(dest, shape, catdims, offsets, x)
+
+* Custom implementations:
+
+    Base.copy(cat::Concatenated{Interface}) # custom implementation of concatenate
+    Base.copyto!(dest, cat::Concatenated{Interface}) # custom implementation of concatenate! based on interface
+    Base.copyto!(dest, cat::Concatenated{Nothing}) # custom implementation of concatenate! based on typeof(dest)
+"""
+module Concatenate
+
+using Compat: @compat
+
+export concatenate, concatenate!
+@compat public Concatenated, cat_offset!, cat_offset1!, copy_or_fill!
+
+using Base: promote_eltypeof
+using .DerivableInterfaces: AbstractInterface, interface
+
+# TODO: named like this because Catted looks so ugly and Cat is module name?
+# cfr Broadcast - Broadcasted
+"""
+    Concatenated{Interface,Dims,Args<:Tuple}
+
+Lazy representation of the concatenation of various `Args` along `Dims`, in order to provide
+hooks to customize the implementation.
+"""
+struct Concatenated{Interface,Dims,Args<:Tuple}
+  interface::Interface
+  dims::Val{Dims}
+  args::Args
+
+  function Concatenated(
+    interface::Union{Nothing,AbstractInterface}, dims::Val{Dims}, args::Tuple
+  ) where {Dims}
+    return new{typeof(interface),Dims,typeof(args)}(interface, dims, args)
+  end
+  function Concatenated(dims, args::Tuple)
+    return Concatenated(interface(args...), dims, args)
+  end
+  function Concatenated{Interface}(dims, args) where {Interface}
+    return Concatenated(Interface(), dims, args)
+  end
+  function Concatenated{Interface,Dims}(args) where {Interface,Dims}
+    return new{Interface,Dims,typeof(args)}(Interface(), Val(Dims), args)
+  end
+end
+
+dims(::Concatenated{A,D}) where {A,D} = D
+DerivableInterfaces.interface(cat::Concatenated) = cat.interface
+
+concatenated(args...; dims) = Concatenated(args, Val(dims))
+
+# allocating the destination container
+# ------------------------------------
+Base.similar(cat::Concatenated) = similar(cat, promote_eltypeof(cat.args...))
+Base.similar(cat::Concatenated, ::Type{T}) where {T} = similar(cat, T, axes(cat))
+function Base.similar(cat::Concatenated, ::Type{T}, ax) where {T}
+  return cat_similar(interface(cat), T, ax)
```

Comment:
Where does this call out to?

Reply:
Maybe we should define:
```julia
Base.similar(interface::AbstractArrayInterface, T::Type, ax)
```
as a general customization point for allocating according to an interface.

That is kind of the role of:
https://github.com/ITensor/DerivableInterfaces.jl/blob/94591bd4ee2b43fa320b1c2df0b214f23e88b22b/src/abstractarrayinterface.jl#L17
https://github.com/ITensor/DerivableInterfaces.jl/blob/94591bd4ee2b43fa320b1c2df0b214f23e88b22b/src/abstractarrayinterface.jl#L88

Reply:
At that point I would wonder if `cat_similar` is needed, since then we can consider the customization points to be:
```julia
function Base.similar(cat::Concatenated, ::Type{T}, ax) where {T}
  return similar(interface(cat), T, ax)
end
```
where you can either overload `Base.similar(cat::Concatenated, elt::Type, ax)`, where you can take into account the interface type and arguments stored in `Concatenated`, or you can let it fall back to determining the output from the general definition of `Base.similar(interface::AbstractArrayInterface, elt::Type, ax)`.

Reply:
The current implementation (mostly placeholder since I was lazy) just calls out to the base one, which more or less calls similar on the first argument (which is what we wanted to avoid in the first place).
I like your suggestions, the only thing I would be a bit careful about is the `arraytype(interface, T)`, since I would expect that we really want to axes information there as well.
So, let me put in your last suggestion, and we can have the overall entry point indeed as `Base.similar(interface, T, ax)`.

Reply:
Agreed, let's get rid of `arraytype(interface, T)` and just go with `Base.similar(interface, T, ax)`. Unfortunately I think I overloaded `arraytype` in one or more downstream packages so getting rid of it will be a breaking change but that should be easy to fix.